umbra/umbra_64.c

```diff
@@ -1261,6 +1283,28 @@ umbra_insert_app_to_shadow_arch(void *drcontext,
                                 reg_id_t *scratch_regs,
                                 int num_scratch_regs)
 {
+#if defined(AARCH64)
+    reg_id_t tmp = *scratch_regs;
+    instru_insert_mov_pc(drcontext, ilist, where, opnd_create_reg(tmp),
+                                OPND_CREATE_INT64(map->disp));
+    PRE(ilist, where, INSTR_CREATE_add(drcontext,
+                                       opnd_create_reg(reg_addr), opnd_create_reg(reg_addr),
+                                       opnd_create_reg(tmp)));
+    if (map->options.scale == UMBRA_MAP_SCALE_UP_2X) {
+        PRE(ilist, where, INSTR_CREATE_lsl(drcontext,
+                                           opnd_create_reg(reg_addr), opnd_create_reg(reg_addr),
+                                           OPND_CREATE_INT8(map->shift)));
+    } else if (map->options.scale <= UMBRA_MAP_SCALE_DOWN_2X) {
+        PRE(ilist, where, XINST_CREATE_slr_s(drcontext,
+                                           opnd_create_reg(reg_addr),
+                                           OPND_CREATE_INT8(map->shift)));
+    }
+    instru_insert_mov_pc(drcontext, ilist, where, opnd_create_reg(tmp),
+                             OPND_CREATE_INT64(map->mask));
+    PRE(ilist, where, INSTR_CREATE_and(drcontext,
```

Comment:
Why did you place the `and` instruction last here, in contrast to the existing code below?

Reply:
yes I have done:
add => shift => and

rather than the existing:
and => add => shift

My understanding is that the purpose is the shift it into some other distinct space range with the correct scale and the order of operations didn't matter too much as long as 'a' mapping was achieved.

AArch64 doesn't use the same segmentation scheme as x86 as far as I know.

Therefore I am using the segment mask variable to simply mask off the lower 2 bits of the address in order to secure a valid address (as commented), and this makes sense to be done as the final operation.

Reply:
> yes I have done:
> add => shift => and
> 
> rather than the existing:
> and => add => shift
> 
> My understanding is that the purpose of the mapping is to shift and scale into some other distinct space range with the correct scale. The order of operations doesn't matter too much as long as 'a' mapping was achieved.
> 

I think it can make a difference. The `and` is used to mask the address, the `add` is used as a displacement value to reach a shadow region, and the `shift` is intended for scaling. While it may not make a difference in some cases, having the `and` last might clear information injected earlier by the displacement/scaling if we make future changes.

In any case, it is much better to have the same code sequence as other platforms if possible, as otherwise it appears we are doing something special and potentially arises confusion. 

> AArch64 doesn't use the same segmentation scheme as x86 as far as I know.
> 
> Therefore I am using the mask variable to simply mask off the lower 2 bits of the address in order to secure a valid address (as commented), and this makes sense to be done as the final operation.

The mask and displacement values are critically set up in order to have shadow memory regions not collide with app regions. These values were carefully crafted, and some recently generated with the help of automated tools (https://github.com/DynamoRIO/drmemory/blob/master/tools/umbra_layout.py).

Maybe I am missing something, but I still don't understand why we can't use the existing segment mapping for Linux?